src/mono/mono/metadata/assembly.c

```diff
@@ -1470,44 +1471,46 @@ bundled_assembly_match (const char *bundled_name, const char *name)
 }
 
 static MonoImage *
-open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status, gboolean is_satellite)
+open_from_bundle_internal (MonoAssemblyLoadContext *alc, const char *filename, MonoImageOpenStatus *status)
 {
-	if (!bundles)
+	if (!bundles && !mono_bundled_resource_get_bundle_contains_assemblies ())
 		return NULL;
 
-	MonoImage *image = NULL;
-	char *name = is_satellite ? g_strdup (filename) : g_path_get_basename (filename);
-	for (int i = 0; !image && bundles [i]; ++i) {
-		if (bundled_assembly_match (bundles[i]->name, name)) {
+	MonoBundledAssemblyResource *assembly = (MonoBundledAssemblyResource *)mono_bundled_resource_get_bundled_resource (filename);
```

Comment:
should we have some kind of assertion that the thing we get back really is a `MonoBundledAsemblyResource` and not some other sort of `MonoBundledResource`?

Reply:
Agree, maybe we should have different getters for the different resource type, that includes an assert + returning the correct type right away, like

```
static inline MonoBundledAssemblyResource *
mono_bundled_resource_get_bundled_assembly_resource (const char *name)
{
    MonoBundledAssemblyResource *assembly =  (MonoBundledAssemblyResource*)mono_bundled_resource_get_bundled_resource (filename);
    g_assert (assembly.type == MONO_BUNDLED_ASSEMBLY);
    return assembly;
}
```

Reply:
I'm not sure we need more functions - maybe just an extra arg of the expected resource type?

Reply:
Would be nice to remove the need to cast to different types on callsite and improve type safety, will just be 3 small additional inlined functions.

Reply:
Inlined 3 getter functions into the header files (`bundled-resources-internals.h` and `mono-bundled-source.h`)